### ISSUE TITLE
nginx config in docker example has wrong rewrite

### ISSUE THREAD

**Issue by mauritsvanrees** [plone/documentation#1335]
### Plone Documentation versions affected

- [ ] 5.x
- [X] 6.x

### Description

TLDR: a rewrite url in this [example nginx config](https://6.dev-docs.plone.org/install/containers/examples/nginx-volto-plone-zeo.html#nginx-configuration) should change.

I started adding an issue in Volto, but later I realised this is a documentation issue.

**Describe the bug**
I setup Docker with nginx, backend, frontend and zeo, as described in the Plone 6 dev docs. I can login directly in the backend, but not in Volto or nginx.
Not sure if my problem is caused by Volto or plone.restapi, docker images, documentation, the nginx config, or something else.

**To Reproduce**

1. Follow the instructions from the example at https://6.dev-docs.plone.org/install/containers/examples/nginx-volto-plone-zeo.html, creating a default.conf and docker compose file.
2. Edit `default.conf` to use `localhost` instead of `plone.localhost`, or add `plone.localhost` to your `/etc/hosts` file.
3. Edit the docker compose file to use `plone/plone-frontend:16.0.0-alpha.26` for the frontend, instead of `latest` (which would be version 15).
4. Run `docker compose up -d` to build and start everything.
5. Go directly to the backend at http://localhost:8080/Plone and login with admin/admin. This works.
6. Go directly to the frontend at http://localhost:3000/. This displays the basic Plone layout but says "Connection refused" because the backend is not available. This is probably expected in this setup, for example it tries to get the breadcrumbs at http://localhost:3000/++api++/@breadcrumbs which won't work.
7. Go to the nginx webserver at port 80, so simply http://localhost/.
8. Try to login with admin/admin. This fails.

At first I wondered if this might be because recently the minimum password length in Plone was increased from 5 to 8. But logging in to the backend works. Ah, and that change is not yet in beta 2 of the backend. So this is not the issue.

According to my browser, the corresponding `curl` command is this:

```
curl 'http://localhost/++api++/@login' -X POST -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0' -H 'Accept: application/json' -H 'Accept-Language: nl,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br' -H 'Content-Type: application/json' -H 'Origin: http://localhost' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Referer: http://localhost/login?return_url=' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache' --data-raw '{"login":"admin","password":"admin"}'
```

Ignoring the boring parts, it boils down to this, which seems correct:

```
curl 'http://localhost/++api++/@login' -X POST -H 'Accept: application/json' --data-raw '{"login":"admin","password":"admin"}'
```

`docker compose logs` says this:

```
py76be-backend-1    | 2022-09-28 22:59:00 ERROR [Zope.SiteErrorLog:35][waitress-0] BadRequest: http://localhost/++api++/POST_application_json_
py76be-backend-1    | Traceback (innermost last):
py76be-backend-1    |   Module ZPublisher.WSGIPublisher, line 167, in transaction_pubevents
py76be-backend-1    |   Module ZPublisher.WSGIPublisher, line 376, in publish_module
py76be-backend-1    |   Module ZPublisher.WSGIPublisher, line 271, in publish
py76be-backend-1    |   Module ZPublisher.mapply, line 85, in mapply
py76be-backend-1    |   Module ZPublisher.WSGIPublisher, line 68, in call_object
py76be-backend-1    |   Module plone.rest.service, line 22, in __call__
py76be-backend-1    |   Module plone.restapi.services, line 19, in render
py76be-backend-1    |   Module plone.restapi.services.content.add, line 41, in reply
py76be-backend-1    | zExceptions.BadRequest: Property '@type' is required
py76be-backend-1    | 172.18.0.5 - - [28/Sep/2022:22:59:00 +0000] "POST /VirtualHostBase/http/localhost/Plone/%2B%2Bapi%2B%2B/VirtualHostRoot/%2B%2Bapi%2B%2B HTTP/1.0" 400 1282 "http://localhost/login?return_url=" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0"
py76be-webserver-1  | 172.18.0.1 - - [28/Sep/2022:22:59:00 +0000] "POST /++api++/@login HTTP/1.1" 400 1282 "http://localhost/login?return_url=" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0" "-"
```

So somehow the `++api++/login` does not make it to the correct login backend endpoint, because it ends up in the [content add](https://github.com/plone/plone.restapi/blob/8.26.0/src/plone/restapi/services/content/add.py#L41) endpoint.
Ah, and if I look closely now, the url that the backend gets, has the escaped version of `++api++` twice. 

For reference, these services are running in my setup:

```
$ docker compose ps
NAME                 COMMAND                  SERVICE             STATUS              PORTS
py76be-backend-1     "/app/docker-entrypo…"   backend             running (healthy)   0.0.0.0:8080->8080/tcp
py76be-db-1          "/app/start-zeo.sh"      db                  running             0.0.0.0:8100->8100/tcp
py76be-frontend-1    "docker-entrypoint.s…"   frontend            running             0.0.0.0:3000->3000/tcp
py76be-webserver-1   "/docker-entrypoint.…"   webserver           running             0.0.0.0:80->80/tcp
```

Also, for the record, I am fairly new to Docker. :-)

**Expected behavior**
Login at port 80 works

**Software (please complete the following information):**

- Volto Version 16.0.0-alpha.26 (assuming this is what `plone/plone-frontend:16.0.0-alpha.26` contains
- Plone 6.0.0b2
- Plone REST API 8.26.0

**Possible solution:**

The rewrite rule is currently this:

      rewrite ^/(\+\+api\+\+\/?)+($|/.*) /VirtualHostBase/http/$server_name/Plone/++api++/VirtualHostRoot/$1 break;

Problem is that the `$1` points to `(\+\+api\+\+\/?)`.

Solution 1: replace `$1` with `$2` because that matches `($|/.*)` so in our example this matches `/@login`. This solves the immediate problem with a one-character fix.

Solution 2: change the first part to something like `rewrite ^/\+\+api\+\+\/($|/.*)`. But I have not tried this yet. And I wonder if the current strange spelling is deliberate in that it would likely match `++api++/++api++/` as the `$1`, which I guess could accidentally occur as url if something is not quite right. But it seems a bit unlikely.

**Comment by stevepiercy** [comment]
To help with testing and explaining regex, I use this nifty tool. This example is already configured with the current regex.

https://regex101.com/r/j0GUnH/1

Any of these would match:

```
/++api++
/++api++/
/++api++/@login
/++api++/++api++
/++api++/++api++/
/++api++/++api++/@login
/++api++/++api++/++api++
/++api++/++api++/++api++/
/++api++/++api++/++api++/@login
```

And for the proposed regex in Solution 2.

https://regex101.com/r/8JgTdm/1

Only this would match:

```
/++api++/
```

I don't think we want that.

I think the original regex is correct. I think that we want to match a string:

- beginning with `/`
- in the first capture group, we match the following and discard it:
  - `++api++` at least one time
  - optionally followed by a trailing `/`
- in the second capture group, match either:
  - end of string `$`, or
  - `/` followed by any number of characters

When logging in, I think we want to pass the second capture group as the `return_url` value in the redirect. I think your analysis of changing `$1` to `$2` is correct. To verify, does this line change by including a value in the POST URL?

```
py76be-backend-1    | 172.18.0.5 - - [28/Sep/2022:22:59:00 +0000] "POST /VirtualHostBase/http/localhost/Plone/%2B%2Bapi%2B%2B/VirtualHostRoot/%2B%2Bapi%2B%2B HTTP/1.0" 400 1282 "http://localhost/login?return_url=" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0"
```

to this:

```
py76be-backend-1    | 172.18.0.5 - - [28/Sep/2022:22:59:00 +0000] "POST /VirtualHostBase/http/localhost/Plone/%2B%2Bapi%2B%2B/VirtualHostRoot/%2B%2Bapi%2B%2B/@login HTTP/1.0" 400 1282 "http://localhost/login?return_url=" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:105.0) Gecko/20100101 Firefox/105.0"
```

?

Side note, I've been using Docker for 6 years, and it is still new to me.

**Comment by Querela** [comment]
I also stumbled over this issue...
For me the above steps worked _(with the templates from the [docs](https://6.dev-docs.plone.org/install/containers/examples/nginx-volto-plone.html))_:
```yml
# docker-compose.yml
frontend:
    image: plone/plone-frontend:16.0.0-alpha.42
```
```nginx
# default.conf
server {
  location ~ /\+\+api\+\+($|/.*) {
      rewrite ^/(\+\+api\+\+\/?)+($|/.*) /VirtualHostBase/http/$server_name/Plone/++api++/VirtualHostRoot/$2 break;
```

I also noticed when trying to specify the latest frontend image version that `latest, 15, 15.16, 15.16.0` for the [plone-frontend](https://github.com/plone/plone-frontend/) does not seem to be correct. It did not find `15.16, 15.16.0`, and `latest` seemed to point to `15.15`? With `latest` the frontend was a bit broken, e.g. `Unknown Block slate` UI elements.

But to fix the login issue the `$1` &rarr; `$2` in the nginx `default.conf` file is necessary.
And then to fix the frontend the `16.0.0-alpha...` version seems to be required.

**Comment by stevepiercy** [comment]
Thanks for the confirmation @Querela. It looks like all the examples that use nginx need to be updated accordingly.

> I also noticed when trying to specify the latest frontend image version that `latest, 15, 15.16, 15.16.0` for the [plone-frontend](https://github.com/plone/plone-frontend/) does not seem to be correct.

It looks like the tags for 15.16 were not pushed to Docker Hub: https://hub.docker.com/r/plone/plone-frontend/tags. I was about to create a new issue, but it looks like someone already reported it: https://github.com/plone/plone-frontend/issues/24.

**Comment by stevepiercy** [comment]
@Querela pull request created at https://github.com/plone/documentation/pull/1348
